### ISSUE TITLE
Add C++17 support

### ISSUE THREAD

**Issue by kazk** [Codewars/codewars-runner-cli#618]
Need [Clang 5](https://clang.llvm.org/cxx_status.html).

**Comment by oschonrock** [comment]
+1

C++14 is 5 yrs old and we are nearing c++20. It's frustrating to have to code in "legacy" style

What version of clang are you running that C++17 is not an option?

**Comment by kazk** [comment]
> What version of clang are you running that C++17 is not an option?

C++ still uses super old Clang 3.8 :(

There's also some problems with using abandoned test framework [Igloo](https://github.com/joakimkarlsson/igloo). I'm not sure how well it'll work with C++17 (#658).
I'd like to eventually use [Catch2](https://github.com/catchorg/Catch2) instead, but will need to support both until existing kata are migrated (or let C++17 kata use Catch2 and older ones use Igloo).

I haven't spent much time on this yet since we haven't had much demand.

We already have Clang 8 for C, so I'll start experimenting with updating C++ soon.

**Comment by hobovsky** [comment]
`-O2` please?

**Comment by kazk** [comment]
Deployed C++17 support. Forgot to add `-O2`, but I can add it later. It'll probably make the compile time slightly longer, but it shouldn't be too bad. Will do some testing.

Please help with updating 57 incompatible kata. See [List of C++ Kata to Update to C++17](https://github.com/Codewars/codewars.com/wiki/List-of-Cpp-Kata-to-Update-to-Cpp17).

**Comment by hobovsky** [comment]
I tried to update two or three kata from "needs update" list (for example, "https://www.codewars.com/kata/multiplying-numbers-as-strings") and it turns out that they suffer from timeouts, because execution times for old C++14 seem to be ~60% of execution time after switching to C++17. I am not sure which part takes longer, compilation, container setup or tests themselves, but it appears that C++17 is almost two times slower.

**Comment by kazk** [comment]
Maybe that's because C++17 one is using `libc++` which is not the default on Linux. It was just easier for me to install the correct version from https://apt.llvm.org/ along with Clang.

`libstdc++` is the default on Linux and C++14 uses it. I'll look into which version of `libstdc++` is available and see how the performance is.

**Comment by kazk** [comment]
I got Clang 8 with `libstdc++-9` working locally. There's many more incompatible kata now, but it's much faster I think. At least "Multiplying numbers as strings" doesn't timeout.

I'll have to re-run the script to validate C++17 compatibility and update the list after deploying this.

**Comment by kazk** [comment]
C++17 with `clang-8` + `libstdc++-9` is deployed and the performance issue seems fixed. I also added `-O2`.

There's many more incompatible kata so some of them with C++17 enabled might not be compatible anymore. I'll re-run the script to validate compatibility soon to fix this.

**Comment by kazk** [comment]
Please help with updating 173 incompatible kata. See [List of C++ Kata to Update to C++17](https://github.com/Codewars/codewars.com/wiki/List-of-Cpp-Kata-to-Update-to-Cpp17).

I'll try looking into common causes and group them if I can. I saw many are missing `#include <chrono>` and using `std::mt19973` inproperly. Those somehow worked when using `libc++`.

**Comment by FArekkusu** [comment]
This is weird. I've just looked into [this kata](https://www.codewars.com/kata/549ee8b47111a81214000941) which is incompatible due to a time out under. It takes ~6 seconds to test 1336 inputs under C++14, and 8 seconds to test 136(!) inputs under C++17. I tried to [measure time](https://stackoverflow.com/a/27739925) spent on every line of code in the tests, and it took at most a couple milliseconds to 1) create a list of pregenerated inputs/outputs; 2) shuffle them; 3) test them all. But at the same time the execution time jumped to 12(!) seconds, and didn't go back after I removed the time measurement code.
